aten/src/ATen/native/quantized/cpu/qnnpack/src/q8dwconv/mp8x25-neon-per-channel.c

```diff
@@ -0,0 +1,939 @@
+        vaccX0_lo =
+            vmlal_s16(vaccX0_lo, vget_low_s16(vxk6), vget_low_s16(vxi6));
+        vaccX0_hi =
+            vmlal_s16(vaccX0_hi, vget_high_s16(vxk6), vget_high_s16(vxi6));
+
+        const uint8x8_t vk7 = vld1_u8(w);
+        w = (void*)((uintptr_t)w + sizeof(uint8x8_t));
+        const uint8x8_t vi7 = vreinterpret_u8_u64(
+            vshl_u64(vreinterpret_u64_u8(vld1_u8(i7)), vi_shift));
+        const int16x8_t vxk7 =
+            vreinterpretq_s16_u16(vsubl_u8(vk7, vkernel_zero_point));
+        const int16x8_t vxi7 =
+            vreinterpretq_s16_u16(vsubl_u8(vi7, vinput_zero_point));
+        vaccX1_lo =
+            vmlal_s16(vaccX1_lo, vget_low_s16(vxk7), vget_low_s16(vxi7));
+        vaccX1_hi =
+            vmlal_s16(vaccX1_hi, vget_high_s16(vxk7), vget_high_s16(vxi7));
+
+        const uint8x8_t vk8 = vld1_u8(w);
+        w = (void*)((uintptr_t)w + sizeof(uint8x8_t));
+        const uint8x8_t vi8 = vreinterpret_u8_u64(
+            vshl_u64(vreinterpret_u64_u8(vld1_u8(i8)), vi_shift));
+        const int16x8_t vxk8 =
+            vreinterpretq_s16_u16(vsubl_u8(vk8, vkernel_zero_point));
+        const int16x8_t vxi8 =
+            vreinterpretq_s16_u16(vsubl_u8(vi8, vinput_zero_point));
+        vaccX0_lo =
+            vmlal_s16(vaccX0_lo, vget_low_s16(vxk8), vget_low_s16(vxi8));
+        vaccX0_hi =
+            vmlal_s16(vaccX0_hi, vget_high_s16(vxk8), vget_high_s16(vxi8));
+
+        const uint8x8_t vk9 = vld1_u8(w);
+        w = (void*)((uintptr_t)w + sizeof(uint8x8_t));
+        const uint8x8_t vi9 = vreinterpret_u8_u64(
+            vshl_u64(vreinterpret_u64_u8(vld1_u8(i9)), vi_shift));
+        const int16x8_t vxk9 =
+            vreinterpretq_s16_u16(vsubl_u8(vk9, vkernel_zero_point));
+        const int16x8_t vxi9 =
+            vreinterpretq_s16_u16(vsubl_u8(vi9, vinput_zero_point));
+        vaccX1_lo =
+            vmlal_s16(vaccX1_lo, vget_low_s16(vxk9), vget_low_s16(vxi9));
+        vaccX1_hi =
+            vmlal_s16(vaccX1_hi, vget_high_s16(vxk9), vget_high_s16(vxi9));
+
+        int32x4_t vacc_lo = vaddq_s32(vaccX0_lo, vaccX1_lo);
+        int32x4_t vacc_hi = vaddq_s32(vaccX0_hi, vaccX1_hi);
+
+        vst1q_s32(outacc, vacc_lo);
+        outacc += 4;
+        vst1q_s32(outacc, vacc_hi);
+        outacc += 4;
+      }
+    }
+    {
+      const uint8_t* i0 = input[10];
+      const uint8_t* i1 = input[11];
+      const uint8_t* i2 = input[12];
+      const uint8_t* i3 = input[13];
+      const uint8_t* i4 = input[14];
+      const uint8_t* i5 = input[15];
+      const uint8_t* i6 = input[16];
+      const uint8_t* i7 = input[17];
+      const uint8_t* i8 = input[18];
+      const uint8_t* i9 = input[19];
+      output = output_start;
+      outacc = outacc32;
+
+      size_t c = channels;
+      for (; c >= 8; c -= 8) {
+        const uint8x8_t vkernel_zero_point =
+            vld1_u8(&quantization_params->neon.kernel_zero_points[channels - c]);
```

Comment:
ARM has very good support for free indexing as part of the memory operations as we have seen in the assembly so I am not sure whether this will be a win, but one common practice in optimization is to aggressively factor out invariables out of loops - in this case:

```
const uint8_t* const kernel_channel_zero_points = quantization_params->neon.kernel_zero_points + channels;

for ( ... ) {
  vld1_u8(kernel_channel_zero_points - c);
   ...
}
```

Reply:
@AshkanAliabadi so do you feel that compiler cannot optimize this? I am assuming you dont actually end up accessing the params and then offset into it to get to kernel_zero_point and then do indexing on it every time. If that is the case I can see that this will be inefficient, but things like this which are loop invariant, at least base pointer `kernel_zero_point`, should be moved outside the loop. I dont have much experience in terms of how efficient compilers are for such things, but thats what I am relying on.
And yes without looking at disassembly you wouldn't know if compiler is able to move it out or not.

Reply:
Anyways, if you prefer I can make that change.

Reply:
I generally don't trust the compiler to do the right thing but that might be an outdated line of thinking.  I think the code is perfectly fine as is.  Generally speaking all my comments are suggestions of possible alternate implementations which may or may not be advantageous. Please feel free to pick and choose as you see fit.

Reply:
These are simple enough changes, so I made them anyway.